consensus/protocol_upgrade_table.go

```diff
@@ -0,0 +1,68 @@
+package consensus
+
+import (
+	"sort"
+
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/pkg/errors"
+)
+
+// protocolUpgrade specifies that a particular protocol version goes into effect at a particular block height
+type protocolUpgrade struct {
+	Version     uint64
+	EffectiveAt *types.BlockHeight
+}
+
+// ProtocolUpgradeTable is a data structure capable of specifying which protocol versions are active at which block heights
+type ProtocolUpgradeTable struct {
+	Network  string
+	upgrades []protocolUpgrade
+}
+
+// NewProtocolUpgradeTable creates a new ProtocolUpgradeTable that only tracks upgrades for the given network
+func NewProtocolUpgradeTable(network string) *ProtocolUpgradeTable {
+	return &ProtocolUpgradeTable{
+		Network: network,
+	}
+}
+
+// Add configures an upgrade for a network. If the network doesn't match the PUT's network, this upgrade will be ignored.
+func (put *ProtocolUpgradeTable) Add(network string, version uint64, effectiveAt *types.BlockHeight) {
+	// ignore upgrade if not part of our network
+	if network != put.Network {
+		return
+	}
```

Comment:
I think I see how this is working with the static upgrade definitions, but it's obscure. The network name is known when the upgrade table is configured and a node only ever needs one table: the one for its network name. Hence the network name should not required as a property on this object at all.

`protocol_upgrades.go` may have a map of network names to lists of protocol versions, but only one entry of that map need be looked up when this table is built.

Reply:
The network name is now only configured in the builder. The table only contains the upgrades.

Reply:
This silent ignoring of the network name still puts me off. I think it would be cleaner if the network parameter was removed.